core/services/blockhashstore/feeder.go

```diff
@@ -114,3 +132,114 @@ func (f *Feeder) Run(ctx context.Context) error {
 	f.lastRunBlock = latestBlock
 	return errs
 }
+
+func (f *Feeder) runTrusted(
+	ctx context.Context,
+	latestBlock uint64,
+	fromBlock uint64,
+	blockToRequests map[uint64]map[string]struct{},
+) error {
+	var errs error
+
+	// Iterate through each request block via waitGroup.
+	// For blocks with pending requests, add them to the batch to be stored.
+	// Note: Golang maps sort items in a range randomly, so although the batch size is used
+	// to limit blocks-per-batch, every block has an equal chance of getting picked up
+	// on each run.
+	var batch = make(map[uint64]struct{})
+	for blockKey, unfulfilledReqs := range blockToRequests {
+		f.wgStored.Add(1)
+		var unfulfilled = unfulfilledReqs
+		var block = blockKey
+		go func() {
+			defer f.wgStored.Done()
+			if len(unfulfilled) == 0 {
+				return
+			}
+			if _, ok := f.stored[block]; ok {
+				// Already stored
+				return
+			}
+
+			// Do not store a block if it has been marked as stored; otherwise, store it even
+			// if the RPC call errors, as to be conservative.
+			stored, err := f.bhs.IsStored(ctx, block)
+			if err != nil {
+				f.lggr.Errorw("Failed to check if block is already stored, attempting to store anyway",
+					"err", err,
+					"block", block)
+				errs = multierr.Append(errs, errors.Wrap(err, "checking if stored"))
```

Comment:
Is this multithreaded write threadsafe?

Reply:
Uhhh, not sure, but this is existing behavior in BHS anyway. Worth an external investigation? Are multierrors not threadsafe?

Reply:
To this and the below, just to be safe I have added locks around all writes and reads of things beyond the scope of the goroutine.